src/modules/aggsig/main_impl.h

```diff
@@ -0,0 +1,303 @@
+
+    /* finalize */
+    secp256k1_scalar_get_b32(partial->data, &sec);
+    secp256k1_scalar_clear(&sec);
+    aggctx->progress[index] = NONCE_PROGRESS_SIGNED;
+    return 1;
+}
+
+int secp256k1_aggsig_combine_signatures(const secp256k1_context* ctx, secp256k1_aggsig_context* aggctx, unsigned char *sig64, const secp256k1_aggsig_partial_signature *partial, size_t n_sigs) {
+    size_t i;
+    secp256k1_scalar s;
+    secp256k1_ge final;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(aggctx != NULL);
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(partial != NULL);
+    (void) ctx;
+
+    if (n_sigs != aggctx->n_sigs) {
+        return 0;
+    }
+
+    secp256k1_scalar_set_int(&s, 0);
+    for (i = 0; i < n_sigs; i++) {
+        secp256k1_scalar tmp;
+        int overflow;
+        secp256k1_scalar_set_b32(&tmp, partial[i].data, &overflow);
+        if (overflow) {
+            return 0;
+        }
+        secp256k1_scalar_add(&s, &s, &tmp);
+    }
+
+    /* If we need to negate the public nonce, everyone will
+     * have negated their secret nonces in the previous step. */
+    if (!secp256k1_gej_has_quad_y_var(&aggctx->pubnonce_sum)) {
+        secp256k1_gej_neg(&aggctx->pubnonce_sum, &aggctx->pubnonce_sum);
+    }
+
+    secp256k1_scalar_get_b32(sig64, &s);
+    secp256k1_ge_set_gej(&final, &aggctx->pubnonce_sum);
+    secp256k1_fe_normalize_var(&final.x);
+    secp256k1_fe_get_b32(sig64 + 32, &final.x);
+    return 1;
+}
+
+
+typedef struct {
+    const secp256k1_context *ctx;
+    unsigned char prehash[32];
+    const secp256k1_pubkey *pubkeys;
+} secp256k1_verify_callback_data;
+
+static int secp256k1_aggsig_verify_callback(secp256k1_scalar *sc, secp256k1_ge *pt, size_t idx, void *data) {
+    secp256k1_verify_callback_data *cbdata = (secp256k1_verify_callback_data*) data;
+
+    if (secp256k1_compute_sighash(sc, cbdata->prehash, idx) == 0) {
+        return 0;
+    }
+    secp256k1_scalar_negate(sc, sc);
+    secp256k1_pubkey_load(cbdata->ctx, pt, &cbdata->pubkeys[idx]);
+    return 1;
+}
+
+int secp256k1_aggsig_verify(const secp256k1_context* ctx, secp256k1_scratch_space *scratch, const unsigned char *sig64, const unsigned char *msg32, const secp256k1_pubkey *pubkeys, size_t n_pubkeys) {
+    secp256k1_scalar g_sc;
+    secp256k1_gej pk_sum;
+    secp256k1_fe fe_tmp;
+    secp256k1_ge r_ge;
+    int overflow;
+    secp256k1_verify_callback_data cbdata;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+    ARG_CHECK(scratch != NULL);
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(pubkeys != NULL);
+    (void) ctx;
+
+    if (n_pubkeys == 0) {
+        return 0;
+    }
+
+    /* extract s */
+    secp256k1_scalar_set_b32(&g_sc, sig64, &overflow);
+    if (overflow) {
+        return 0;
+    }
+
+    /* extract R */
+    if (!secp256k1_fe_set_b32(&fe_tmp, sig64 + 32)) {
+        return 0;
+    }
+    if (!secp256k1_ge_set_xquad(&r_ge, &fe_tmp)) {
```

Comment:
When n_pubkeys is 1 (which is probably going to be pretty common), there is a more efficient verification algorithm where you don't decompress R, but instead just compare the X coordinate of the recomputed version, and check that its Y coordinate is a quadratic residue.

Reply:
Nice. I don't think this even needs `n_pubkeys` to be 1.